### PR TITLE
Adding distroless image to Envoy CI pipeline

### PR DESCRIPTION
**Pull request by oleksiyp**
Commit Message: ci: adding distroless image to Envoy CI pipeline
Additional Description: 

In #16170 we came up to the conclusion that the best option is to add distroless based images.
I analyzed #15312 and came to the conclusion that it will take a lot of changes to add distroless images by using bazel's rules_docker. Instead of it using current CI pipeline to add distroless image is low hanging fruit. Conversion to rules_docker might be performed later.

Risk Level: Medium
Testing: Please guide what testing best to perform
Docs Changes: Please guide if there are any best location to document it
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Comment by phlax** [comment]
cc @lizan

**Comment by phlax** [comment]
once this lands i will to through the examples and start shifting over anything that currently uses the alpine images

we also need to update docs to include some refs to this image

**Comment by oleksiyp** [comment]
Hold on a little bit. I am still trying to run it locally. Wanted to trigger CI build to check it and open PR for discussion. Will return back with a commit or acknowledgment that everything is okay.

**Comment by phlax** [comment]
cool, yep

it seems to be failing with a missing `/bin/sh` on the `RUN` line - not sure how this is done with distroless

**Comment by phlax** [comment]
i think you need

```
RUN ["/busybox/sh", "-c", "mkdir /opt/myfolder"]
```

from https://github.com/GoogleContainerTools/distroless/issues/419#issuecomment-542342659

**Comment by oleksiyp** [comment]
Don't think sh is included. Only in ":debug" images

**Comment by phlax** [comment]
this might work alternatively

```
FROM gcr.io/distroless/java:debug
SHELL ["/busybox/sh", "-c"]
RUN mkdir /opt/myfolder
```

i think `sh` is there in busybox - i guess the key is to prevent it invoking the normal path in `RUN`

**Comment by phlax** [comment]
hmm - maybe not - testing this locally and busybox isnt there either

**Comment by oleksiyp** [comment]
Shell is added only to ":debug" images + 'python' distroless has "dash" because there is no other way to run it.

**Comment by phlax** [comment]
i think then maybe copying in the directory using a multistage build

**Comment by phlax** [comment]
something vaguely like this https://github.com/GoogleContainerTools/distroless/issues/427#issue-514498948

**Comment by oleksiyp** [comment]
I think RUN is not needed here at all.
> The ADD destination is automatically created if it doesn't exist.
https://www.reddit.com/r/docker/comments/7t5953/dockerfile_add_to_a_nonexisting_directory/

**Comment by phlax** [comment]
relatedly - this is the getenvoy Dockerfile - https://github.com/tetratelabs/getenvoy/blob/master/Dockerfile

not sure exactly why they use the `:cc` image but it looks like it has shell support

File: ci/Dockerfile-envoy-distroless [hunk 1]
```diff
@@ -0,0 +1,10 @@
+FROM gcr.io/distroless/base-debian10:nonroot
+
+ADD configs/envoyproxy_io_proxy.yaml /etc/envoy/envoy.yaml
+
+ARG ENVOY_BINARY_SUFFIX=_stripped
+ADD linux/amd64/build_release${ENVOY_BINARY_SUFFIX}/* /usr/local/bin/
```
Comment (phlax) [inline ci/Dockerfile-envoy-distroless:6]: given the lack of `sh` etc im  wondering if `$PATH` is set 

might be necessary to hardcode the path to the binary

Comment (oleksiyp) [reply ci/Dockerfile-envoy-distroless:6]: Path is set. Checking layout of image through ":debug" image. The problem is libc version. I managed to build it locally by building via bazel envoy-static + packaging it via Dockerfile-envoy-distroless and it fails on:

```
oleksiyp@pop-os:~/dev/envoy$ docker run a7d14c019784
envoy: /lib/x86_64-linux-gnu/libm.so.6: version `GLIBC_2.29' not found (required by envoy)
envoy: /lib/x86_64-linux-gnu/libpthread.so.0: version `GLIBC_2.30' not found (required by envoy)
envoy: /lib/x86_64-linux-gnu/libc.so.6: version `GLIBC_2.32' not found (required by envoy)
```
But maybe it captures my libc from system.

Comment (oleksiyp) [reply ci/Dockerfile-envoy-distroless:6]: Image `distroless-debian9` contains:
```
libc-2.24.so
libm-2.24.so
libpthread-2.24.so
```
while `debian10`:
```
libc-2.28.so
libm-2.28.so
libpthread-2.28.so
```

**Comment by oleksiyp** [comment]
> relatedly - this is the getenvoy Dockerfile - https://github.com/tetratelabs/getenvoy/blob/master/Dockerfile
> 
> not sure exactly why they use the `:cc` image but it looks like it has shell support

```
RUN ["/getenvoy", "fetch", "@"]
```
and
```
RUN mkdir  bla-bla-bla
```

differs that second type of command runs shell I believe, while first just uses `exec` syscall.

**Comment by phlax** [comment]
i see

so iiuc this should work now 

i had a poke around the debian and cc images - not sure what is extra in the cc image but its slightly larger - so lets use the debian image if we can

**Comment by phlax** [comment]
as its a pita to test the ci images locally i created a testing branch here #16269

File: ci/verify_examples.sh [hunk 2]
```diff
@@ -32,6 +32,7 @@ run_examples () {
     examples=$(find . -mindepth 1 -maxdepth 1 -type d -name "$TESTFILTER" ! -iname "_*" ! -name "$TESTEXCLUDES" | sort)
     for example in $examples; do
         pushd "$example" > /dev/null || return 1
+        chmod go+r -R . || return 1
```
Comment (phlax) [inline ci/verify_examples.sh:35]: yep - this was the other way - this is done for a reason

Comment (phlax) [reply ci/verify_examples.sh:35]: or rather - there is another place where the opposite of this is done - to mimick the env of users with strict umask settings - so we could temp removed that

File: examples/dynamic-config-cp/Dockerfile-proxy [hunk 3]
```diff
@@ -3,4 +3,4 @@ FROM envoyproxy/envoy-dev:latest
 RUN apt-get -y update && apt-get install -y -qq --no-install-recommends netcat
 COPY ./envoy.yaml /etc/envoy.yaml
 RUN chmod go+r /etc/envoy.yaml
-CMD ["/usr/local/bin/envoy", "-c /etc/envoy.yaml", "-l", "debug"]
+CMD ["/usr/local/bin/envoy", "-c /etc/envoy.yaml", "--bootstrap-version", "3", "-l", "debug"]
```
Comment (phlax) [inline examples/dynamic-config-cp/Dockerfile-proxy:6]: this is unrelated - i was testing something else

**Comment by oleksiyp** [comment]
Should I try other examples? all of them?

**Comment by phlax** [comment]
> Should I try other examples? all of them?

no that is good i think

if you revert all of the example stuff im happy for this to merge

if you wish we could try the static distroless base image first

**Comment by oleksiyp** [comment]
Reverted

**Comment by oleksiyp** [comment]
Forgot about `static`. Indeed let's check. Running in parallel #16272

**Comment by oleksiyp** [comment]
From my side, all set for merge.

**Review by phlax — APPROVED** [review]
@oleksiyp thanks for taking the time to work through and test this

lgtm

**Comment by oleksiyp** [comment]
@phlax thank you for your assistance! Envoyproxy is a really cool thing. I hope will have more opportunities to contribute.

Not able to merge:
```
The base branch restricts merging to authorized users
```

**Comment by oleksiyp** [comment]
@phlax  seems build failed.
Some authorization is required:
```
#8 ERROR: server message: insufficient_scope: authorization failed
------
 > exporting to image:
------
error: failed to solve: rpc error: code = Unknown desc = server message: insufficient_scope: authorization failed
The push refers to repository [docker.io/envoyproxy/envoy-distroless-dev]
21c123484da7: Preparing
f3e1a44f8191: Preparing
5d09c2db1d76: Preparing
417cb9b79ade: Preparing
denied: requested access to the resource is denied
```

**Comment by phlax** [comment]
eek!

ive pinged @lizan in the envoy-ci channel on slack - i reckon he will know how to resolve

**Comment by phlax** [comment]
@oleksiyp dockerhub should be fixed now - the first image has been uploaded - you can see it here https://hub.docker.com/r/envoyproxy/envoy-distroless-dev/tags?page=1&ordering=last_updated

i opened a couple of follow up bugs - for documentation (#16282 ), and deprecation of the alpine image here (#16275 ) which is basically fixing the examples that (mistakenly) use the alpine image as a base for backend service
